gcsfs/tests/test_core.py

```diff
@@ -929,3 +929,66 @@ def test_validate_response():
     md5 = repr(base64.b64encode(hashlib.md5(b"foo").digest()))[2:-1]
     with pytest.raises(ChecksumError):
         gcs.validate_response(0, b"f", None, "/path", {"X-Goog-Hash": f"md5={md5}"})
+
+
+@my_vcr.use_cassette(match=["all"])
+@pytest.mark.parametrize(
+    ["file_path", "validate_get_error", "validate_list_error", "expected_error"],
+    [
+        (
+            "/missing",
+            FileNotFoundError,
+            None,  # Not called
+            FileNotFoundError,
+        ),
+        (
+            "/missing",
+            OSError("Forbidden"),
+            FileNotFoundError,
+            FileNotFoundError,
+        ),
+        (
+            "/2014-01-01.csv",
+            None,
+            None,
+            None,
+        ),
+        (
+            "/2014-01-01.csv",
+            OSError("Forbidden"),
+            None,
+            None,
+        ),
+    ],
+    ids=[
+        "missing_with_get_perms",
+        "missing_with_list_perms",
+        "existing_with_get_perms",
+        "existing_with_list_perms",
+    ],
+)
+def test_metadata_read_permissions(
```

Comment:
Don't love how roundabout/messy this test is, so open for other ideas. 😁

Reply:
Don't worry about messy tests! It's when the main code is messy that we are in trouble.